Added mapping documentation into es exporter readme (#34014)
open-telemetry/opentelemetry-collector-contrib: exporter/elasticsearchexporter/README.md

[data stream]: https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html
[ecs]: https://www.elastic.co/guide/en/ecs/current/index.html
[SemConv]: https://github.com/open-telemetry/semantic-conventions

> ## ECS Mapping

Reviewer: Should this be linking to semantic conventions?

Reviewer: should this conversion be in the semantic-conventions specification?

Author: i don't think so

Reviewer: I think it should. Here is the equivalent for log mapping for Splunk HEC: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model-appendix.md#splunk-hec

Author: i see, there's also an ECS specific mapping also defined there: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model-appendix.md#elastic-common-schema

do you suggest adding a link here pointing to that mapping

Author: added a backlink to this readme

When `mode` is set to `ecs`, `elasticsearchexporter` performs conversions for resource-level attributes from their Semantic Conventions (SemConv) names to equivalent Elastic Common Schema (ECS) names.

If the target ECS field name is specified as an empty string (""), the converter will neither convert the SemConv key to the equivalent ECS name nor pass through the SemConv key as-is to become the ECS name.

> "Preserved" means the convention should be preserved in ECS mode. This can happen when an attribute needs to be mapped to an ECS equivalent but also preserved in its original form.

Reviewer: Do you mean that for `host.name`, we will duplicate the attribute to `host.hostname`?

Is this treatment only true for the `host.name` attribute? Can this be handled as a separate paragraph?

Author: so far it is just for host.hostname.
i went with this format just in case we have more exceptions later

Reviewer: the column header says "Preserve" but you use the word "Preserved". If this is exceptional treatment, maybe it needs to be called out more. I also don't understand exactly what you write. Do you mean this?
```suggestion
When "Preserved" is true, the attribute will be preserved in the payload and duplicated as mapped to its ECS equivalent.
```

| host.name                | host.hostname | true |
| k8s.deployment.name      | kubernetes.deployment.name | false |
| k8s.namespace.name       | kubernetes.namespace | false |
| k8s.node.name            | kubernetes.node.name | false |
| k8s.pod.name             | kubernetes.pod.name | false |
| k8s.pod.uid              | kubernetes.pod.uid | false |
| os.description           | host.os.full | false |
| os.name                  | host.os.name | false |
| os.type                  | host.os.platform | false |
| os.version               | host.os.version | false |
| process.executable.path  | process.executable | false |
| process.runtime.name     | service.runtime.name | false |
| process.runtime.version  | service.runtime.version | false |
| service.instance.id      | service.node.name | false |
| telemetry.distro.name    | "" | false |
| telemetry.distro.version | "" | false |
| telemetry.sdk.language   | "" | false |
| telemetry.sdk.name       | "" | false |
| telemetry.sdk.version    | "" | false |

### Compound Mapping

There are ECS fields that are not mapped easily 1 to 1 but require more advanced logic.

#### `agent.name`

> The agent name takes the form of a compound name consisting of 3 components:

Reviewer: Can those component values contain the character `/` ?

Author: i dont see why not, but having `telemetry.sdk.language/name/version` none of them has this so far

Reviewer: It'd be great that it doesn't happen in the future too.

- `telemetry.sdk.name` or, if not present, defaults to `otlp`,
- `telemetry.sdk.language`, defaulting to `unknown` in case it is missing,
- `telemetry.distro.name`, which is allowed to be empty.

These values are all valid:

| `telemetry.sdk.name` | `telemetry.sdk.language` | `telemetry.distro.name` | `agent.name`           |
|----------------------|--------------------------|-------------------------|------------------------|

> | ""                   | ""                       | ""                      | `otlp`                 |

Reviewer: Should that not be `otlp/unknown` ?

| ""                   | dotnet                   | ""                      | `otlp/dotnet`          |
| opentelemetry        | dotnet                   | ""                      | `opentelemetry/dotnet` |
| ""                   | java                     | parts-unlimited-java    | `otlp/java/parts-unlimited-java` |
| ""                   | ""                       | parts-unlimited-java    | `otlp/unknown/parts-unlimited-java` |

#### `agent.version`

> Takes the value of `telemetry.distro.version` or `telemetry.sdk.version`.

Reviewer: Is there a precedence?

Author: yes in the order i mentioned them. do you think we could make it more clear/explicit?

Author: made it explicit
